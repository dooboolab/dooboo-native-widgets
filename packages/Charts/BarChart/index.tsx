import * as d3 from 'd3';

import { G, Line, Rect, Svg, Text } from 'react-native-svg';
import React, { FC } from 'react';

import { BarChartProps } from './types';
import { Hoverable } from 'react-native-web-hooks';
import styled from 'styled-components/native';

// Styled component declaration
const Container = styled.View`
  height: 100%;
  width: 100%;
  background-color: transparent;
  justify-content: center;
  align-items: center;
`;
const HeaderContainer = styled.View`
  width: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
`;
const ChartContainer = styled.View`
  height: 100%;
  width: 100%;
  flex: 1;
  transform: scale(1, -1);
`;

const BarChart: FC<BarChartProps> = (props) => {
  // Destructuring props & Declaring default values
  const {
    /* ====== [REQUIRED] ====== */
    data = [],
    xAxisKey = 'key3',
    yAxisKey = 'key6',
    yUnit = '10',
    /* ====== [OPTIONAL] ====== */
    header = undefined,
    graphStyle = {
      barWidth: 30,
      color: '#000000',
      strokeWidth: 0,
      strokeColor: 'rgba(0,0,0,0.1)',
      withText: true,
      textColor: '#000000',
      textStrokeColor: 'none',
      fontSize: '12',
      fontWeight: 'bold',
    },
    /* **** [Y-AXIS PARAMETERS] **** */
    yStyle = {
      withLabel: true,
      withIndicator: true,
      withLine: true,
      lineColor: '#000000',
      lineStrokeWidth: 1,
      withText: true,
      textColor: '#000000',
      textStrokeColor: 'none',
      fontSize: '12',
      fontWeight: 'normal',
    },
    /* **** [X-AXIS PARAMETERS] ====== **** */
    xStyle = {
      withLabel: true,
      withIndicator: true,
      withLine: true,
      lineColor: '#000000',
      lineStrokeWidth: 1,
      withText: true,
      textColor: '#000000',
      textStrokeColor: 'none',
      fontSize: '12',
      fontWeight: 'bold',
    },
  } = props;

  // Initialize
  const [currentHeight, setCurrentHeight] = React.useState<number>(400);
  const [currentWidth, setCurrentWidth] = React.useState<number>(500);

  // get Y axis values range
  const getYmaxRange = (
    data: Array<Record<string, any>>,
    yUnit: string | number,
  ): Array<number> => {
    const range: number[] = [];
    const unit = typeof yUnit === 'string' ? Number(yUnit) : yUnit;
    const yMaxValue =
      Math.trunc(
        Math.max(...data.map((current) => current[yAxisKey]), 0) / unit,
      ) *
        unit +
      unit;
    for (let i = 0; i <= yMaxValue; i += unit) {
      i <= yMaxValue && range.push(i);
    }
    return range;
  };

  // SVG Layout
  const SVGPadding = 60; // default padding: 30
  const SVGHeight = currentHeight;
  const SVGWidth = currentWidth;
  const indicatorSize = 5;
  const textGap = 10;

  // X-AXIS
  const xDomain = data.map((item, index) => index);
  const xRange = [SVGPadding, SVGWidth - SVGPadding];
  const xAxis = d3.scalePoint().domain(xDomain).range(xRange);

  // Y-AXIS
  const yAxisRange = getYmaxRange(data, yUnit);
  const yDomain = [0, d3.max(yAxisRange, (item) => item)];
  const yRange = [
    SVGPadding,
    SVGHeight - SVGPadding * (SVGHeight < 736 ? 1.5 : SVGHeight < 1023 ? 2.5 : 3.5),
  ];
  const yAxis = d3.scaleLinear().domain(yDomain).range(yRange);
  return (
    <Container
      onLayout={(e): void => {
        setCurrentHeight(e.nativeEvent.layout.height);
        setCurrentWidth(e.nativeEvent.layout.width);
      }}>
      {/* Header */}
      {header && <HeaderContainer>{header}</HeaderContainer>}
      {/* Graph view */}
      <ChartContainer>
        <Svg
          height={SVGHeight}
          width={SVGWidth}
          preserveAspectRatio="xMidYMid slice">
          {/* Graph Y-axis labels view */}
          <G>
            {yStyle.withLine && (
              <Line
                x1={xAxis(0)}
                y1={yAxis(0)}
                x2={xAxis(0)}
                y2={yAxis(yAxisRange[yAxisRange.length - 1])}
                stroke={yStyle.lineColor}
                strokeWidth={yStyle.lineStrokeWidth}
              />
            )}
            {yStyle.withLabel &&
              yAxisRange.map((unit, index) => {
                return (
                  <G key={index}>
                    <Line
                      x1={xAxis(0)}
                      y1={yAxis(unit)}
                      x2={xAxis(data.length - 1) + Number(graphStyle.barWidth)}
                      y2={yAxis(unit)}
                      stroke={'rgba(0,0,0, 0.2)'}
                      strokeDasharray={'2'}
                      strokeWidth={yStyle.lineStrokeWidth}
                    />
                    {yStyle.withIndicator && (
                      <Line
                        x1={xAxis(0) - indicatorSize}
                        y1={yAxis(unit)}
                        x2={xAxis(0)}
                        y2={yAxis(unit)}
                        stroke={yStyle.lineColor}
                        strokeWidth={yStyle.lineStrokeWidth}
                      />
                    )}
                    {yStyle.withText && (
                      <Text
                        scale={[1, -1]}
                        fill={yStyle.textColor}
                        stroke={yStyle.textStrokeColor}
                        fontSize={yStyle.fontSize}
                        fontWeight={yStyle.fontWeight}
                        x={xAxis(0) - indicatorSize - textGap}
                        y={-yAxis(unit) + 4}
                        textAnchor="end">
                        {unit}
                      </Text>
                    )}
                  </G>
                );
              })}
            {/* Graph X-axis labels view */}
            {xStyle.withLabel && xStyle.withLine && (
              <Line
                x1={xAxis(0)}
                x2={
                  xAxis(data.length - 1) +
                  (typeof graphStyle.barWidth === 'string'
                    ? Number(graphStyle.barWidth)
                    : graphStyle.barWidth)
                }
                y1={yAxis(0)}
                y2={yAxis(0)}
                stroke={xStyle.lineColor}
                strokeWidth={xStyle.lineStrokeWidth}
              />
            )}
            {xStyle.withLabel &&
              data.map((item, index) => {
                return (
                  <G key={index}>
                    {xStyle.withIndicator && (
                      <Line
                        x1={
                          xAxis(index) +
                          (typeof graphStyle.barWidth === 'string'
                            ? Number(graphStyle.barWidth)
                            : graphStyle.barWidth) /
                            2
                        }
                        y1={yAxis(0)}
                        x2={
                          xAxis(index) +
                          (typeof graphStyle.barWidth === 'string'
                            ? Number(graphStyle.barWidth)
                            : graphStyle.barWidth) /
                            2
                        }
                        y2={yAxis(0) - indicatorSize}
                        stroke={xStyle.lineColor}
                        strokeWidth={xStyle.lineStrokeWidth}
                      />
                    )}
                    {xStyle.withText && (
                      <Text
                        scale={[1, -1]}
                        fill={xStyle.textColor}
                        stroke={xStyle.textStrokeColor}
                        fontSize={xStyle.fontSize}
                        fontWeight={xStyle.fontWeight}
                        x={
                          xAxis(index) +
                          (typeof graphStyle.barWidth === 'string'
                            ? Number(graphStyle.barWidth)
                            : graphStyle.barWidth) /
                            2
                        }
                        y={-yAxis(0) + 5 + indicatorSize + textGap}
                        textAnchor={'middle'}>
                        {item[xAxisKey]}
                      </Text>
                    )}
                  </G>
                );
              })}
            {/* Graph: Text, Dots & Line */}
            {data.map((item, index) => {
              return (
                <G key={index}>
                  <Hoverable>
                    {(isHovered): React.ReactElement => (
                      <Rect
                        x={xAxis(index)}
                        y={yAxis(0)}
                        width={graphStyle.barWidth}
                        height={yAxis(item[yAxisKey]) - SVGPadding}
                        fill={graphStyle.color}
                        opacity={isHovered ? 0.5 : 0.8}
                        strokeWidth={graphStyle.strokeWidth}
                        stroke={graphStyle.strokeColor}
                        ry={1}
                      />
                    )}
                  </Hoverable>
                  <Text
                    scale={[1, -1]}
                    fill={item[yAxisKey] === 0 ? 'red' : graphStyle.textColor}
                    stroke={graphStyle.textStrokeColor}
                    fontSize={graphStyle.fontSize}
                    fontWeight={graphStyle.fontWeight}
                    x={
                      xAxis(index) +
                      (typeof graphStyle.barWidth === 'string'
                        ? Number(graphStyle.barWidth)
                        : graphStyle.barWidth) /
                        2
                    }
                    y={-yAxis(item[yAxisKey]) - textGap}
                    textAnchor="middle">
                    {item[yAxisKey]}
                  </Text>
                </G>
              );
            })}
          </G>
        </Svg>
      </ChartContainer>
    </Container>
  );
};

export default BarChart;
