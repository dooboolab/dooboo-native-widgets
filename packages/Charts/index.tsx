import BarChart from './BarChart';
import LineChart from './LineChart';

export { LineChart, BarChart };
