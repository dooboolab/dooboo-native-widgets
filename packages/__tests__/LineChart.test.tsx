import * as React from 'react';

import {
  RenderResult,
  render,
} from '@testing-library/react-native';

import { LineChart } from '../Charts';

let props: any;
let component: React.ReactElement;
let testingLib: RenderResult;

const testingData = [
  {
    id: 'abcd1234efgh5674',
    key1: 'D-6',
    key2: Math.trunc(Math.random() * 10),
    key3: '06/30',
    key4: Math.trunc(Math.random() * 1000),
    key5: 'A',
    key6: Math.trunc(Math.random() * 10000),
  },
  {
    id: 'abcd1234efgh5675',
    key1: 'D-5',
    key2: Math.trunc(Math.random() * 10),
    key3: '06/29',
    key4: Math.trunc(Math.random() * 1000),
    key5: 'B',
    key6: Math.trunc(Math.random() * 10000),
  },
  {
    id: 'abcd1234efgh5676',
    key1: 'D-4',
    key2: Math.trunc(Math.random() * 10),
    key3: '06/28',
    key4: Math.trunc(Math.random() * 1000),
    key5: 'C',
    key6: Math.trunc(Math.random() * 10000),
  },
  {
    id: 'abcd1234efgh5677',
    key1: 'D-3',
    key2: Math.trunc(Math.random() * 10),
    key3: '06/27',
    key4: Math.trunc(Math.random() * 1000),
    key5: 'D',
    key6: Math.trunc(Math.random() * 10000),
  },
  {
    id: 'abcd1234efgh5678',
    key1: 'D-2',
    key2: Math.trunc(Math.random() * 10),
    key3: '06/26',
    key4: Math.trunc(Math.random() * 1000),
    key5: 'E',
    key6: Math.trunc(Math.random() * 10000),
  },
  {
    id: 'abcd1234efgh5677',
    key1: 'D-1',
    key2: Math.trunc(Math.random() * 10),
    key3: '06/25',
    key4: Math.trunc(Math.random() * 1000),
    key5: 'F',
    key6: Math.trunc(Math.random() * 10000),
  },
  {
    id: 'abcd1234efgh5678',
    key1: 'D-day',
    key2: Math.trunc(Math.random() * 10),
    key3: '06/24',
    key4: Math.trunc(Math.random() * 1000),
    key5: 'G',
    key6: Math.trunc(Math.random() * 10000),
  },
];

const createTestProps = (
  obj?: Record<string, unknown>,
): Record<string, unknown> => ({
  ...obj,
});

describe('[LineChart] renders', () => {
  it('should render without crashing', () => {
    props = createTestProps({
      data: testingData,
      xAxisKey: 'key5',
      yAxisKey: 'key2',
      yUnit: '2',
    });
    component = <LineChart {...props} />;
    testingLib = render(component);
    expect(testingLib.baseElement).toBeTruthy();
  });

  it('should render when custom props given', () => {
    jest.useFakeTimers();

    props = createTestProps({
      data: testingData,
      xAxisKey: 'key1',
      yAxisKey: 'key4',
      yUnit: '100',
    });
    component = <LineChart {...props} />;

    testingLib = render(component);
    jest.runAllTimers();
    expect(testingLib.baseElement).toBeTruthy();
  });
});
