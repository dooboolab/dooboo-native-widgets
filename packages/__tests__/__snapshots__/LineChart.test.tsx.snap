// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[LineChart] renders should render when custom props given 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "flexBasis": 0,
            "flexDirection": "column",
            "flexGrow": 1,
            "flexShrink": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexDirection": "row",
              "flexGrow": 1,
              "flexShrink": 1,
              "height": "100%",
              "width": "100%",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexDirection": "column",
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <RNSVGSvgView
            bbHeight={400}
            bbWidth={500}
            focusable={false}
            height={400}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 400,
                  "width": 500,
                },
              ]
            }
            transform="scale(1, -1)"
            transform-origin="center"
            width={500}
          >
            <RNSVGGroup
              matrix={
                Array [
                  1,
                  0,
                  0,
                  -1,
                  0,
                  0,
                ]
              }
            >
              <RNSVGGroup
                matrix={
                  Array [
                    1,
                    0,
                    -0,
                    1,
                    30,
                    30,
                  ]
                }
              >
                <RNSVGLine
                  propList={
                    Array [
                      "stroke",
                      "strokeWidth",
                    ]
                  }
                  stroke={4278190080}
                  strokeWidth={1}
                  x1={25}
                  x2={25}
                  y1={NaN}
                  y2={0}
                />
                <RNSVGLine
                  propList={
                    Array [
                      "stroke",
                      "strokeWidth",
                    ]
                  }
                  stroke={4278190080}
                  strokeWidth={1}
                  x1={15}
                  x2={440}
                  y1={10}
                  y2={10}
                />
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={1}
                    x1={50}
                    x2={50}
                    y1={10}
                    y2={5}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        50,
                      ]
                    }
                    y={
                      Array [
                        10,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="6일차"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={1}
                    x1={108.33333333333334}
                    x2={108.33333333333334}
                    y1={10}
                    y2={5}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        108.33333333333334,
                      ]
                    }
                    y={
                      Array [
                        10,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="5일차"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={1}
                    x1={166.66666666666669}
                    x2={166.66666666666669}
                    y1={10}
                    y2={5}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        166.66666666666669,
                      ]
                    }
                    y={
                      Array [
                        10,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="4일차"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={1}
                    x1={225}
                    x2={225}
                    y1={10}
                    y2={5}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        225,
                      ]
                    }
                    y={
                      Array [
                        10,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="3일차"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={1}
                    x1={283.33333333333337}
                    x2={283.33333333333337}
                    y1={10}
                    y2={5}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        283.33333333333337,
                      ]
                    }
                    y={
                      Array [
                        10,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="2일차"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={1}
                    x1={341.6666666666667}
                    x2={341.6666666666667}
                    y1={10}
                    y2={5}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        341.6666666666667,
                      ]
                    }
                    y={
                      Array [
                        10,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="1일차"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={1}
                    x1={400}
                    x2={400}
                    y1={10}
                    y2={5}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        400,
                      ]
                    }
                    y={
                      Array [
                        10,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="금일"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={2}
                    x1={50}
                    x2={108.33333333333334}
                    y1={NaN}
                    y2={NaN}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        50,
                      ]
                    }
                    y={
                      Array [
                        NaN,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="06-30"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                  <RNSVGCircle
                    cx={50}
                    cy={NaN}
                    fill={4294967295}
                    propList={
                      Array [
                        "fill",
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    r="4"
                    stroke={4278190080}
                    strokeWidth={2}
                  />
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={2}
                    x1={108.33333333333334}
                    x2={166.66666666666669}
                    y1={NaN}
                    y2={NaN}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        108.33333333333334,
                      ]
                    }
                    y={
                      Array [
                        NaN,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="06-29"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                  <RNSVGCircle
                    cx={108.33333333333334}
                    cy={NaN}
                    fill={4294967295}
                    propList={
                      Array [
                        "fill",
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    r="4"
                    stroke={4278190080}
                    strokeWidth={2}
                  />
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={2}
                    x1={166.66666666666669}
                    x2={225}
                    y1={NaN}
                    y2={NaN}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        166.66666666666669,
                      ]
                    }
                    y={
                      Array [
                        NaN,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="06-28"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                  <RNSVGCircle
                    cx={166.66666666666669}
                    cy={NaN}
                    fill={4294967295}
                    propList={
                      Array [
                        "fill",
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    r="4"
                    stroke={4278190080}
                    strokeWidth={2}
                  />
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={2}
                    x1={225}
                    x2={283.33333333333337}
                    y1={NaN}
                    y2={NaN}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        225,
                      ]
                    }
                    y={
                      Array [
                        NaN,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="06-27"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                  <RNSVGCircle
                    cx={225}
                    cy={NaN}
                    fill={4294967295}
                    propList={
                      Array [
                        "fill",
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    r="4"
                    stroke={4278190080}
                    strokeWidth={2}
                  />
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={2}
                    x1={283.33333333333337}
                    x2={341.6666666666667}
                    y1={NaN}
                    y2={NaN}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        283.33333333333337,
                      ]
                    }
                    y={
                      Array [
                        NaN,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="06-26"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                  <RNSVGCircle
                    cx={283.33333333333337}
                    cy={NaN}
                    fill={4294967295}
                    propList={
                      Array [
                        "fill",
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    r="4"
                    stroke={4278190080}
                    strokeWidth={2}
                  />
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={2}
                    x1={341.6666666666667}
                    x2={400}
                    y1={NaN}
                    y2={NaN}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        341.6666666666667,
                      ]
                    }
                    y={
                      Array [
                        NaN,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="06-25"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                  <RNSVGCircle
                    cx={341.6666666666667}
                    cy={NaN}
                    fill={4294967295}
                    propList={
                      Array [
                        "fill",
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    r="4"
                    stroke={4278190080}
                    strokeWidth={2}
                  />
                </RNSVGGroup>
                <RNSVGGroup>
                  <RNSVGLine
                    propList={
                      Array [
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    stroke={4278190080}
                    strokeWidth={2}
                    x1={400}
                    x2={400}
                    y1={NaN}
                    y2={NaN}
                  />
                  <RNSVGText
                    content={null}
                    dx={Array []}
                    dy={Array []}
                    fill={4278190080}
                    font={
                      Object {
                        "fontSize": "12",
                        "fontWeight": "bold",
                        "textAnchor": "middle",
                      }
                    }
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        -1,
                        0,
                        0,
                      ]
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    rotate={Array []}
                    stroke={null}
                    x={
                      Array [
                        400,
                      ]
                    }
                    y={
                      Array [
                        NaN,
                      ]
                    }
                  >
                    <RNSVGTSpan
                      content="06-24"
                      dx={Array []}
                      dy={Array []}
                      font={Object {}}
                      rotate={Array []}
                      x={Array []}
                      y={Array []}
                    />
                  </RNSVGText>
                  <RNSVGCircle
                    cx={400}
                    cy={NaN}
                    fill={4294967295}
                    propList={
                      Array [
                        "fill",
                        "stroke",
                        "strokeWidth",
                      ]
                    }
                    r="4"
                    stroke={4278190080}
                    strokeWidth={2}
                  />
                </RNSVGGroup>
              </RNSVGGroup>
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
  </View>
</View>
`;
